### PR TITLE
[SPARK-41596][SS][DOCS] Document the new feature "Async Progress Tracking" to Structured Streaming guide doc

### PR DESCRIPTION
**Pull request by jerrypeng**
### What changes were proposed in this pull request?

Add documentation on how async progress tracking should be used

### Why are the changes needed?

Documentation on async progress tracking is missing

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?

n/a

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Doesn't look like to be ready to review, assuming PR description, missing picture, etc. I'll add [WIP] right now.

**Comment by jerrypeng** [comment]
@HeartSaVioR Can you take another look please?

File: docs/structured-streaming-programming-guide.md [hunk 1]
```diff
@@ -3569,7 +3569,63 @@ the effect of the change is not well-defined. For all of them:
     structures into bytes using an encoding/decoding scheme that supports schema migration. For example,
     if you save your state as Avro-encoded bytes, then you are free to change the Avro-state-schema between query
     restarts as the binary state will always be restored successfully.
+# Asynchronous Progress Tracking
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:3572]: nit: one empty line

File: docs/structured-streaming-programming-guide.md [hunk 2]
```diff
@@ -3569,7 +3569,63 @@ the effect of the change is not well-defined. For all of them:
...
+  
+## What is it?
+
+Asynchronous progress tracking allows Structured Streaming pipelines to checkpoint progress asynchronously and in parallel to the actual data processing within a micro-batch, reducing latency associated with maintaining the offsetLog and commitLog.
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:3576]: nit: pipelines -> queries, maybe `Structured Streaming pipelines` -> `streaming queries`

Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md:3576]: nit: offsetLog -> offset log, commitLog -> commit log.

File: docs/structured-streaming-programming-guide.md [hunk 3]
```diff
@@ -3569,7 +3569,63 @@ the effect of the change is not well-defined. For all of them:
...
+
+![Async Progress Tracking](img/async-progress.png)
+
+## How does it work?
+
+Structured Streaming relies on persisting and managing offsets as progress indicators for query processing. Offset management operation directly impacts processing latency, because no data processing can occur until these operations are complete. Asynchronous progress tracking enables Structured Streaming pipelines to checkpoint progress without being impacted by these offset management operations.
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:3582]: same: `pipelines` -> `queries`, maybe `Structured Streaming pipelines` -> `streaming queries`

File: docs/structured-streaming-programming-guide.md [hunk 4]
```diff
@@ -3569,7 +3569,63 @@ the effect of the change is not well-defined. For all of them:
...
+
+| Option    | Value           | Default | Description       | 
+|-------------|-----------------|------------|---------------------|
+|asyncProgressTrackingEnabled|true/false|false|enable or disable asynchronous progress tracking|
+|asyncProgressCheckpointingInterval|minutes|1|the interval in which we commit offsets and completion commits|
+
+## Limitations
+The initial version of the feature has the following limitations:
+
+* Asynchronous progress tracking is only supported in  stateless pipelines using Kafka Sink
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:3611]: nit: stateless query, no double spaces

File: docs/structured-streaming-programming-guide.md [hunk 5]
```diff
@@ -3569,7 +3569,63 @@ the effect of the change is not well-defined. For all of them:
...
+* Exactly once end-to-end processing will not be supported with this asynchronous progress tracking  because offset ranges for batch can be changed in case of failure. Though many sinks, such as Kafka sink, do not support writing exactly once anyways.
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:3612]: nit: no double spaces.

File: docs/structured-streaming-programming-guide.md [hunk 6]
```diff
@@ -3569,7 +3569,63 @@ the effect of the change is not well-defined. For all of them:
...
+
+## Switching the setting off
+Turning the async progress tracking off my cause the following exception to be thrown
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:3615]: nit: my -> may

File: docs/structured-streaming-programming-guide.md [hunk 7]
```diff
@@ -3569,7 +3569,63 @@ the effect of the change is not well-defined. For all of them:
...
+```
+
+
+Also the following error message may be printed in the driver logs:
+
+```
+The offset log for batch x doesn't exist, which is required to restart the query from the latest batch x from the offset log. Please ensure there are two subsequent offset logs available for the latest batch via manually deleting the offset file(s). Please also ensure the latest batch for commit log is equal or one batch earlier than the latest batch for offset log.
+```
 
+This is caused by the fact that when async progress tracking is enabled, the framework will not checkpoint progress for every batch as would be done if async progress tracking is not used.  To solve this problem simply re-enable “asyncProgressTrackingEnabled” and set “asyncProgressCheckpointingInterval” to 0 and run the streaming query until at least two micro-batches have been processed.  Async progress tracking can be now safely disabled and restarting query should proceed normally.
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:3628]: nit: no double spaces.

**Review by HeartSaVioR — COMMENTED** [review]
All nits, thanks for working on it!

**Comment by jerrypeng** [comment]
@HeartSaVioR thanks for the review.  PTAL

**Comment by HeartSaVioR** [comment]
스크린샷 2023-01-16 오전 8 15 43

Could you please check there is empty line between normal text vs header? In above, header is indented which shouldn't, because there is no empty line between the two.

File: docs/structured-streaming-programming-guide.md (hunk 1 above)
Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md:3572]: This seems to be missed.

File: docs/structured-streaming-programming-guide.md [hunk 8]
```diff
@@ -3569,7 +3569,62 @@ the effect of the change is not well-defined. For all of them:
...
+```
+
+
+Also the following error message may be printed in the driver logs:
+
+```
+The offset log for batch x doesn't exist, which is required to restart the query from the latest batch x from the offset log. Please ensure there are two subsequent offset logs available for the latest batch via manually deleting the offset file(s). Please also ensure the latest batch for commit log is equal or one batch earlier than the latest batch for offset log.
+```
+
+This is caused by the fact that when async progress tracking is enabled, the framework will not checkpoint progress for every batch as would be done if async progress tracking is not used. To solve this problem simply re-enable “asyncProgressTrackingEnabled” and set “asyncProgressCheckpointingInterval” to 0 and run the streaming query until at least two micro-batches have been processed. Async progress tracking can be now safely disabled and restarting query should proceed normally.
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:3627]: nit: empty line after this line

**Comment by jerrypeng** [comment]
@HeartSaVioR PTAL!  Thank you!

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
스크린샷 2023-01-18 오전 10 28 59

I just verified manually that the page shows the content correctly.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master. (I don't think this needs CI verification as we saw green in prev commit and we just added empty lines.)

**Pull request closed without merging**